### ISSUE TITLE
powerAmpache song sharing 

### ISSUE THREAD

**Issue by potagerGit** [owncloud/music#1167]
With power ampache2 Android app, i would like to share a song via the share button, and it doesn't work.
Maybe not a bug but a feature request ?

**Comment by paulijar** [comment]
Yes, the sharing part of the Ampache API is not implemented at time being.

There is also another feature request https://github.com/owncloud/music/issues/781 about the sharing support for the Subsonic API. If the support would be built for one of the APIs (Subsonic/Ampache), then the other one would probably need very little extra effort to get the similar support.

**Comment by fab-kaz** [comment]
Hello all,

any news for the feature ?

regards,

f.

**Comment by paulijar** [comment]
@fab-kaz Sorry, no news and no concrete plans or schedule for this.

**Comment by fab-kaz** [comment]
hey @paulijar,

thank you for your answer ;)

maybe i could do some stuff, can you point me some links to starts with ?
regards,
f.

**Comment by paulijar** [comment]
@fab-kaz Pull requests are welcome :). 

The Ampache endpoints to implement are `share_create`, `share_delete`, `share_edit`, `share`, and `shares`. You can find the API documentation from here: https://ampache.org/api/api-json-methods/#shares.

To add Ampache endpoints, all you need to do is add the corresponding functions to https://github.com/owncloud/music/blob/master/lib/Controller/AmpacheController.php like this:
```
    /**
     * @AmpacheAPI
     */
    protected function share_create(string $filter, string $type, ?string $description=null, ?int $expires=null): array {
        // do something with the share API of ownCloud/Nextcloud
        // ...
        return [
            "id" => $filter,
            "name" => "Name of the share",
            "owner" => $this->userId(),
            "allow_stream" => true,
            "allow_download" => true,
            ...
        ];
    }
```
The annotation `@AmpacheAPI` there triggers some reflection magic which will bind this function to the corresponding REST API call.

I don't know the Nextcloud share API all that well so that's something to figure out. Anyway, it should be possible to inject the API interface to the AmpacheController by just adding it to the constructor of the class (provided you are using the latest software from the master branch, I have done some related modernization during the past month).

The Ampache API enables sharing many types of entities. I believe that sharing the songs is the most straight forward thing here as the songs are files system objects and sharing them should align well with the existing capabilities of Nextcloud/ownCloud. I don't know what would it take to enable sharing of albums or playlists.

I also installed the OG Ampache 7.7.1 to get some feel on how the shares are actually supposed to work. But for some reason, all my created share links failed to open with some permission error.
